Introduce Controller to Check API Server Health

DirectXMan12 opened the pull request:
The "kubernetes" service (in the "default" namespace) has endpoints which
point to any API servers active in the cluster.  Each API server is in charge
of maintaining the list of endpoints -- it ensures that its own endpoint
is present, and trims the list down to a fixed number of endpoints,
thereby maintaining that the list will eventually contain the correct
number of active API servers

However, conditions may occur in which an API server crashes or
otherwise becomes unreachable, and a new server is not brought online
(or is not brought online in a timely fashion).  In this case, one or
more addresses in the "kubernetes" endpoints may exist which point to
unreachable API servers.  This, in turn, can cause issues (such as
delays) as clients attempt to use the "kubernetes" service, and requests
time out or are not answered.

To remedy this, a new controller is introduced which periodically checks
the health of the API servers using by checking the '/healthz' URL
served by the API servers.  If it finds an unreachable API server, it
removes the corresponding address from the "kubernetes" endpoints
list (if the API server comes back up, it will re-add itself to the
list).  The controller is off by default, but may be turned on by
setting the `--health-check-controller-sync-peroid` to a non-zero
time period.

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/20975)

<!-- Reviewable:end -->

DirectXMan12 commented:
cc @ncdc @kubernetes/rh-cluster-infra

ncdc commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -165,6 +167,7 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.IntVar(&s.TerminatedPodGCThreshold, "terminated-pod-gc-threshold", s.TerminatedPodGCThreshold, "Number of terminated pods that can exist before the terminated pod garbage collector starts deleting terminated pods. If <= 0, the terminated pod garbage collector is disabled.")
 	fs.DurationVar(&s.HorizontalPodAutoscalerSyncPeriod, "horizontal-pod-autoscaler-sync-period", s.HorizontalPodAutoscalerSyncPeriod, "The period for syncing the number of pods in horizontal pod autoscaler.")
 	fs.DurationVar(&s.DeploymentControllerSyncPeriod, "deployment-controller-sync-period", s.DeploymentControllerSyncPeriod, "Period for syncing the deployments.")
+	fs.DurationVar(&s.HealthCheckControllerSyncPeriod, "health-check-controller-sync-peroid", s.HealthCheckControllerSyncPeriod, "Period for checking the the health of all API servers (set to 0 to disable)")
```
typo: peroid

smarterclayton replied:
The name of this needs to actually correspond to what it is doing - "apiserver-health-check-period"

DirectXMan12 replied:
ACK

ncdc commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,159 @@
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+type APIServerHealthChecker interface {
```
Godoc

DirectXMan12 replied:
ACK

smarterclayton commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,159 @@
+		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no TCP port was called 'https'", masterAddress)
+	}
+
+	url := fmt.Sprintf("https://%s:%v/%s", masterAddress, masterPort, c.healthzPath)
```
This needs to be using the Probe classes.

smarterclayton replied:
You can't assume https.

DirectXMan12 replied:
> You can't assume https.

AFAICT, kube only publishes HTTPS port in the endpoints list.  I could also tell it to look for an 'http' endpoint if available.

DirectXMan12 replied:
> This needs to be using the Probe classes

As in `pkg/probe/http`?  IMO, it doesn't look quite like what we want:

- The health checker interface should expect to receive a master IP, not be required to give a full URL (that way, different checkers could implement different checks in the future)

- We want to use the client information supplied to kube (using the right client certs, trusting the right CAs, etc) -- the HTTPProber constructor just constructs a plain HTTP client `InsecureSkipVerify` and `DisableKeepAlives` set to true.

Using the `DoHTTPProbe` function instead of manually running `NewRequest` and `client.Do` seems to make sense, though.

smarterclayton replied:
I definitely don't want two health checkers in our codebase doing subtly
different things.  The core prober and the core health checker codebase
should converge, not diverge.  If that means minor changes to prober, I
think that's appropriate.  Larger refactoring is probably not in the cards,
but I don't want proliferation of health probe-y things

On Wed, Feb 10, 2016 at 1:33 PM, Solly Ross <notifications@github.com>
wrote:

> In pkg/controller/healthzcheck/controller.go
> <https://github.com/kubernetes/kubernetes/pull/20975#discussion_r52501939>
> :
>
> > +		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no ports provided", masterAddress)
> > +	}
> > +
> > +	masterPort := -1
> > +	for _, port := range masterPorts {
> > +		if port.Name == "https" && port.Protocol == api.ProtocolTCP {
> > +			masterPort = port.Port
> > +			break
> > +		}
> > +	}
> > +
> > +	if masterPort == -1 {
> > +		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no TCP port was called 'https'", masterAddress)
> > +	}
> > +
> > +	url := fmt.Sprintf("https://%s:%v/%s", masterAddress, masterPort, c.healthzPath)
>
> This needs to be using the Probe classes
>
> As in pkg/probe/http? IMO, it doesn't look quite like what we want:
>
>    -
>
>    The health checker interface should expect to receive a master IP, not
>    be required to give a full URL (that way, different checkers could
>    implement different checks in the future)
>    -
>
>    We want to use the client information supplied to kube (using the
>    right client certs, trusting the right CAs, etc) -- the HTTPProber
>    constructor just constructs a plain HTTP client InsecureSkipVerify and
>    DisableKeepAlives set to true.
>
> Using the DoHTTPProbe function instead of manually running NewRequest and
> client.Do seems to make sense, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20975/files#r52501939>.
>

smarterclayton commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -350,6 +351,22 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
+
+		healthCheckClient := clientset.NewForConfigOrDie(config)
+		checkController := healthzcheckcontroller.NewAPIServerHealthCheckController(
+			healthCheckClient.Core().Endpoints("default"), healthChecker, "kubernetes", recorder)
```
This assumes that the controller can reach the apiservers on this address - you should note that in the godoc.

DirectXMan12 replied:
ACK

derekwaynecarr replied:
The api server takes a parameter for what namespace has the master kubernetes service.

Do we need to add a new arg to match kube-apiserver:
```
fs.StringVar(&s.MasterServiceNamespace, "master-service-namespace", s.MasterServiceNamespace, "The namespace from which the kubernetes master services should be injected into pods")
```

smarterclayton replied:
Yes, good catch.

aveshagarwal commented on pkg/apis/componentconfig/types.go:
```diff
@@ -471,6 +476,9 @@ type KubeControllerManagerConfiguration struct {
 	LeaderElection LeaderElectionConfiguration `json:"leaderElection"`
 	// vloumeConfiguration holds configuration for volume related features.
 	VolumeConfiguration VolumeConfiguration `json:"volumeConfiguration"`
+	// masterServiceNamespace is The namespace from which the kubernetes
```
typo: The-> the

DirectXMan12 replied:
ack.

aveshagarwal commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,189 @@
+// httpsHealthzChecker checks API server health by issuing a GET request over HTTPS to
+// the API server at the configured "healthz" path (e.g. "/healthz").  In order for this
+// to work, the caller (the controller manager) must be able to reach all of the API servers.
+// It search for HTTP ports if no HTTPS ports are available.
```
typo: search-> searches

DirectXMan12 replied:
ack.

aveshagarwal commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -353,6 +354,26 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
+		// endpoints.
+		checkerTransport, err := client.TransportFor(config)
+		if err != nil {
+			panic(err)
```
Out of curiosity want to understand the reason behind this panic. Why to panic here?

DirectXMan12 replied:
This is essentially the same behavior as the `OrDie` functions used above (e.g. `clientset.NewForConfigOrDie` -- try to construct a new client, and `panic` if something fails, because we can't proceed without such fundamental operation (and probably don't want to proceed with launch).

smarterclayton replied:
We should be using glog.fatal, not panic.

aveshagarwal replied:
I am really thinking why not to continue with a warning as if this feature was not enabled, like defaulting to this-feature-less behavior.

smarterclayton replied:
The user asked for health checking, we should never silently fail.

aveshagarwal replied:
I agree about not doing anything silently. I was suggesting continue with a warning and not failing 
It seems to me that the main controller works if all sub controller work and fails even if one fails. If that is desired/or by design, that's fine with me.

DirectXMan12 replied:
@smarterclayton ACK.  I was just following the style of the `OrDie` commands (since this is effectively `NewHTTPClientOrDie`) but `glog.Fatalf` makes sense too.

aveshagarwal commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,189 @@
+		// use the HTTP port if necessary, but let any HTTPS ports override it
+		if healthPort == -1 && port.Name == "http" && port.Protocol == api.ProtocolTCP {
+			healthPort = port.Port
+			healthProto = "http"
```
Could use continue here to avoid next check.

aveshagarwal commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,189 @@
+		status, body, err := http_probe.DoHTTPProbe(urlParsed, map[string][]string{}, c.client)
+		if err != nil || status != probe.Success {
+			glog.V(5).Infof("Unable to probe health URL for master %s (at %s) on retry %v/%v: error: %v, status: %v, result: %s", masterAddress, urlParsed, retries+1, maxRetries, err, status, body)
+			delay *= 2
```
delay is doubling in the first loop itself, that means 125ms is never used. Fine if that is the intention.

DirectXMan12 replied:
Yep, that's the intention.

akram commented:
@DirectXMan12 @aveshagarwal Hi guys, do you have any follow-up regarding this issue?
could it be rebased ?
Do you see any workaround?

ncdc commented:
@DirectXMan12 hopefully will have time to rebase next week

DirectXMan12 commented:
Whoops, sorry, had the rebase Friday but must have not pushed send properly ;-)

ncdc commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+// APIServerHealthCheckController is a controller loop which peroidically checks each
```
s/loop//

s/peroidically/periodically/

DirectXMan12 replied:
fixed

ncdc commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,188 @@
+	eventRecorder     record.EventRecorder
+}
+
+// NewAPIServerHealthCheckController constructs a APIServerHealthCheckController with the provided endpoints
```
s/a APIServerHealthCheckController/an APIServerHealthCheckController/

DirectXMan12 replied:
fixed

smarterclayton commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,188 @@
+	maxRetries := 4
+	retries := 0
+	delay := 125 * time.Millisecond
+	for retries < maxRetries {
```
Is there a reason this isn't using wait.ExponentialBackoff?  Don't know if I asked that before.

DirectXMan12 replied:
I don't think you asked that before.  I'll take a look (I have to page this back into RAM :-P).

smarterclayton replied:
This may have predated ExponentialBackoff

On Mon, Apr 25, 2016 at 3:49 PM, Solly Ross <notifications@github.com>
wrote:

> In pkg/controller/healthzcheck/controller.go
> <https://github.com/kubernetes/kubernetes/pull/20975#discussion_r60976249>
> :
>
> > +	if len(masterPorts) == 0 {
> > +		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no ports provided", masterAddress)
> > +	}
> > +
> > +	masterPort, masterProto, err := c.choosePort(masterPorts)
> > +	if err != nil {
> > +		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz URL: %v", masterAddress, err)
> > +	}
> > +	urlParsed, err := url.Parse(fmt.Sprintf("%s://%s:%v/%s", masterProto, masterAddress, masterPort, c.healthzPath))
> > +	if err != nil {
> > +		return false, err
> > +	}
> > +	maxRetries := 4
> > +	retries := 0
> > +	delay := 125 * time.Millisecond
> > +	for retries < maxRetries {
>
> I don't think you asked that before. I'll take a look (I have to page this
> back into RAM :-P).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20975/files/05374c7b831ab3f2ceaaff6f8234dce8c408dde4#r60976249>
>

smarterclayton commented:
Looks mostly ok to me, one comment, @kubernetes/kube-api for an API review of the changes to `componentconfigs`

bgrant0607 commented:
cc @fabioy

bgrant0607 commented:
See also the "component registration" PRs and issues: #13216, #18610

cc @lavalamp

bgrant0607 commented:
cc @mikedanese

bgrant0607 commented on pkg/apis/componentconfig/types.go:
```diff
@@ -522,6 +527,9 @@ type KubeControllerManagerConfiguration struct {
 	ControllerStartInterval unversioned.Duration `json:"controllerStartInterval"`
+	// masterServiceNamespace is the namespace from which the kubernetes
+	// master services should be injected into pods.
+	MasterServiceNamespace string `json:"masterServiceNamespace"`
```
We're trying to avoid the term "master". Is this just about the apiserver?

bgrant0607 commented on pkg/apis/componentconfig/types.go:
```diff
@@ -474,6 +474,11 @@ type KubeControllerManagerConfiguration struct {
+	// If <= 0, the API server health checks are disabled.  In order to work, the
+	// controller manager must be able to reach all of the API servers as specified in the
+	// "kubernetes" auto-created endpoints list.
+	APIServerHealthCheckPeriod unversioned.Duration `json:"apiServerHealthCheckPeriod"`
```
"Health check" is ambiguous, which is why we don't use it in the API.

This sounds like a readiness probe. I suggest calling it that instead.

bgrant0607 commented on pkg/controller/healthzcheck/controller.go:
```diff
@@ -0,0 +1,188 @@
+limitations under the License.
+*/
+
+package healthzcheck
```
Let's make the name reasonably consistent with the configuration parameter. 

componentreadiness?

lavalamp commented:
I do not agree with this mechanism. It should be pretty easy to have apiservers negotiate their number in etcd.

lavalamp commented:
Sorry, I did not see this PR until just now. :(

ncdc commented:
@lavalamp could you elaborate?

ncdc commented:
@smarterclayton if we move DNS to the nodes instead of having it in the OpenShift master, we wouldn't need this any more, right?

lavalamp commented:
Each apiserver should create & maintain a key in an etcd directory with a short ttl. Then they compute the number of expected apiservers by counting the keys in that directory.

ncdc commented:
Ok, so you're not against something being able to remove endpoints - you're just against the http probe?

lavalamp commented:
I'm against writing a new controller to do this.

If you don't like my suggestion above, the "correct" option: we already have an endpoint controller that does this. Just make pod objects for every apiserver (already done?), create a real service instance that references them (trivial change) and you'll get this behavior for free. I'm very much against doing something special for apiserver. Just have to be a little careful to leave some limited bootstrapping code in apiserver so we never get into a situation where endpoint controller can't contact apiserver because the endpoints aren't filled out, but that should be a pretty small modification of the existing mechanism.

DirectXMan12 commented:
> If you don't like my suggestion above...

Just to clarify your solution above: each apiserver creates a key in etcd, figures out the equivalent endpoints list, and then updates the list of endpoints if need be?  That seems reasonable to me.

lavalamp commented:
I was intending to just track the number of apiservers, but if you want to make the value of the key provide enough information to populate the endpoints list, I can live with that, too.

cc @mikedanese

DirectXMan12 commented:
> If you don't like my suggestion above, the "correct" option: we already have an endpoint controller that does this. Just make pod objects for every apiserver (already done?), create a real service instance that references them (trivial change) and you'll get this behavior for free

Yeah, that seems like it would be great for the containerized apiserver case, but I suspect it might present some difficulties for the non-containerized case.

If people don't have any objections, I'll fine with moving forward with the "short TTL etcd key" method.

smarterclayton commented:
Let's sketch out "short" though - 1 minute?  30 seconds?

lavalamp commented:
I was thinking in the 5-15 second range. Should be a drop in the bucket compared to our other etcd traffic.

> Yeah, that seems like it would be great for the containerized apiserver case, but I suspect it might present some difficulties for the non-containerized case.

I suppose we shouldn't block on getting people to use containers...

And sorry again for dumping cold water on this so late.

smarterclayton commented:
To refresh the key you'll need to deal with retries, timing windows, and ensuring the create / fail errors are properly handled.  However, how do they know which entries to remove from the list?

DirectXMan12 commented:
> Let's sketch out "short" though - 1 minute? 30 seconds?

Presumably, it would be similar/equivalent to the current EndpointInterval, which defaults to 10 seconds (which seems reasonable to me).

lavalamp commented:
The current code handles removing the right endpoint (eventually). If you
store the endpoint in the etcd endpoint key, then that makes it even easier.

Yes, you want to refresh your key before it actually expires.

On Tue, Apr 26, 2016 at 1:43 PM, Solly Ross <notifications@github.com>
wrote:

> Let's sketch out "short" though - 1 minute? 30 seconds?
>
> Presumably, it would be similar/equivalent to the current
> EndpointInterval, which defaults to 10 seconds (which seems reasonable to
> me).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20975#issuecomment-214879740>
>

DirectXMan12 commented:
> The current code handles removing the right endpoint (eventually)

Presuming that something new comes to take its place (which is part of the problem here).

I do imagine, though, that it would work somewhat similarly to the current functionality, except that instead of blindly trimming down (and only trimming down when something comes up to replace a dead API server), it would use the information in etcd to do the trimming/repopulation.

DirectXMan12 commented:
Also, to respond to @bgrant0607 above:

> See also the "component registration" PRs and issues: #13216, #18610

It does seem like #18610 would be a good eventual solution here, which somewhat mirrors what I have currently in this PR.  However, #18610 seems like it would probably take a longer period of time to put in place (API changes, etc), so we should probably solve the issue here, and then move over to #18610 when it's finished.

DirectXMan12 committed:
Automatically remove inactive api server endpoints

This commit automatically removes API servers that have crashed or
otherwise lost connectivity to etcd from the list of endpoints for
the "kubernetes" service in the "defaut" namespace.

Unlike the previous code, dead endpoints will be removed, even if a
new endpoint does not come up to replace it.  This removes the
requirement on setting the master count on each API server, and
thus the `--apiserver-count` flag has been deprecated, and no longer
has any functionality.

DirectXMan12 commented:
ok, I've update according to @lavalamp's proposal.  There's a new non-public API object used to store each of the IPs in etcd, each with a TTL of `(refresh_interval + 2) seconds)`.  I wasn't sure what the policy was for changing/removing fields from the master config object, so I just marked the flags associated with `MasterCount` as deprecated, but left the fields in place, although they no longer have any effect.

Some of the generated stuff seems to have changed for no apparent reason, so I'll try to figure out what's up there.

DirectXMan12 commented:
@smarterclayton any idea why the codec generator seems to be updating unrelated stuff?

ncdc commented on pkg/master/controller.go:
```diff
@@ -31,21 +31,93 @@ import (
+	})
+}
+
+// SetLeaseTTL configures the base TTL to use when resetting the TTL in UpdateLease
```
SetLeaseTime

ncdc commented on pkg/master/controller.go:
```diff
@@ -31,21 +31,93 @@ import (
+
+// SetLeaseTTL configures the base TTL to use when resetting the TTL in UpdateLease
+func (mgr *storageMasterLeaseManager) SetLeaseTime(ttl uint64) {
+	mgr.leaseTime = ttl + 2 // give ourselves some wiggle room
```
Why the wiggle?

ncdc commented on pkg/master/controller.go:
```diff
@@ -226,19 +302,19 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
+// TTL and the key will expire.  ReconcileEndpoints will notice that the endpoints object
+// is different from the directory listing, and update the endpoints object accordingly.
 func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool) error {
+
```
remove blank line

sttts commented on pkg/api/types.go:
```diff
@@ -2570,6 +2570,25 @@ type RangeAllocation struct {
+// APIServerIPInfo is an opaque API object (not exposed to end users) that is temporarily persisted (with
+// a short TTL) to etcd to indicate that a particular IP is a viable API server.  The list of such objects
+// is used to determine the endpoints for the autogenerated "kubernetes" service.
+type APIServerIPInfo struct {
```
Wouldn't it work to store the apiserver endpoints with this ttl directly?

smarterclayton replied:
I think we want separate objects and keys anyway.  However, we don't need a new API type.  Just store one endpoints object per apiserver.

sttts commented on pkg/master/controller.go:
```diff
@@ -31,21 +31,93 @@ import (
+
+		glog.V(6).Infof("Resetting TTL on master IP %q listed in storage to %v", ip, leaseTime)
+
+		// NB: GuaranteedUpdate helpfull does not perform the set unless something has changed,
```
s/helpfull/helpfully/

sttts commented on pkg/master/controller.go:
```diff
@@ -31,21 +31,93 @@ import (
 )
 
+// MasterLeaseManager is an interface which assists in managing the set of active masters
+type MasterLeaseManager interface {
```
It's called c.MasterLeases lateron. Why not remove this meaningless word "Manager"?

sttts commented on pkg/master/controller.go:
```diff
@@ -250,50 +326,48 @@ func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointP
+		return fmt.Errorf("No master IPs were listed in storage, refusing to erase all endpoints for the kubernetes service")
+	}
+
+	// Next, we compare the the current list of endpoints with the list of master IP keys
```
the the

sttts commented on pkg/master/controller.go:
```diff
@@ -226,19 +302,19 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
-// understand the requirements and the body of this function.
-//
-// Requirements:
-//  * All apiservers MUST use the same ports for their {rw, ro} services.
```
Why is this comment gone? It's still a requirement to have equal ports.

sttts commented on pkg/master/controller.go:
```diff
@@ -250,50 +326,48 @@ func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointP
-		// We *always* add our own IP address.
-		e.Subsets[0].Addresses = append(e.Subsets[0].Addresses, api.EndpointAddress{IP: ip.String()})
+
+	if !formatCorrect || !ipCorrect {
```
Maybe simpler with a `switch` block and `fallthrough` .

sttts commented on pkg/master/controller.go:
```diff
@@ -250,50 +326,48 @@ func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointP
 		// Something is egregiously wrong, just re-make the endpoints record.
 		e.Subsets = []api.EndpointSubset{{
-			Addresses: []api.EndpointAddress{{IP: ip.String()}},
+			Addresses: []api.EndpointAddress{},
```
unnecessary. It's set anyway below.

sttts commented on pkg/master/controller.go:
```diff
@@ -302,11 +376,10 @@ func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointP
 // * formatCorrect is true if exactly one subset is found.
-// * ipCorrect is true when current master's IP is found and the number
-//     of addresses is less than or equal to the master count.
+// * ipCorrect when the addresses in the endpoints match the expected addresses list
```
s/ipCorrect/ipsCorrect/g

sttts commented on pkg/master/controller.go:
```diff
@@ -323,12 +396,30 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
+			presentAddrs[ip] = false
+		}
+
+		// currentAddrs == expectedAddr iff (len(ca) == len(ea) && ca <= ea && ca has no duplicates)
```
confusing comment. We actually don't check the uniqueness, but assume it as a pre-condition.

sttts commented on pkg/master/controller.go:
```diff
@@ -226,19 +302,19 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
+		return err
+	}
+
+	// Then, retrive the current list of endpoints...
```
retrieve

sttts commented on pkg/master/controller.go:
```diff
@@ -226,19 +302,19 @@ func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, ser
+// is different from the directory listing, and update the endpoints object accordingly.
 func (c *Controller) ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool) error {
+
+	// First, refresh the TTL on our key
```
I would highlight the point that "First" means that we update our ttl in every iteration, independently whether any error or update conflict happens below. This makes sure that at least some of the master will add our endpoint.

smarterclayton commented on pkg/master/controller.go:
```diff
@@ -31,21 +31,93 @@ import (
 	"k8s.io/kubernetes/pkg/util/wait"
 )
 
+// MasterLeaseManager is an interface which assists in managing the set of active masters
```
Move this to its own package.

smarterclayton commented on pkg/master/controller.go:
```diff
@@ -67,7 +139,8 @@ type Controller struct {
 
-	runner *util.Runner
+	runner       *util.Runner
+	MasterLeases MasterLeaseManager
```
Don't put public vars in the same block as private vars, move this up.

smarterclayton commented:
If you remove the new API object and make this opt-in I think this is something we can merge for 1.3 *if* the old code path is preserved.  How hard is it to leave the exact old code path and have the new opt in path?  That would allow us to merge this and get soak time in origin before we flip it on for Kube.

smarterclayton commented:
And by old code path being preserved, I mean by observation it's obvious that old code path is called in app places if the lease manager is nil.

smarterclayton commented:
Superseded by #26574 and #26915
